xds/src/main/java/io/grpc/xds/LookasideLb.java

```diff
@@ -248,6 +196,215 @@ private static ManagedChannel initLbChannel(
+      XdsClient xdsClient;
+      @Nullable
+      LocalityStore localityStore;
+      @Nullable
+      LoadReportClient lrsClient;
+
+      FixedXdsConfigBalancer(Helper helper) {
+        this.helper = helper;
+      }
+
+      @Override
+      public void handleNameResolutionError(Status error) {}
+
+      @Override
+      public boolean canHandleEmptyAddressListFromNameResolution() {
+        return true;
+      }
+
+      @Override
+      public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+        if (xdsClientRef != null) {
+          // handleResolvedAddresses() already called, so no-op here, because the config is fixed in
+          // this balancer.
+          return;
+        }
+
+        // TODO(zdapeng): First try to get loadStatsStore from XdsAttributes.LOAD_STATS_STORE_REF.
+        LoadStatsStore loadStatsStore = new LoadStatsStoreImpl();
+        localityStore = localityStoreFactory.newLocalityStore(
+            helper, lbRegistry, loadStatsStore);
+        // TODO(zdapeng): Use XdsClient to do Lrs directly.
+        final LoadReportCallback lrsCallback =
+            new LoadReportCallback() {
+              @Override
+              public void onReportResponse(long reportIntervalNano) {
+                localityStore.updateOobMetricsReportInterval(reportIntervalNano);
+              }
+            };
+
+        // There are three usecases:
+        // 1. The EDS-only legacy usecase that requires balancerName from xds config.
+        //    TODO(zdapeng): Remove the legacy case.
+        // 2. The EDS-only with bootstrap usecase:
+        //    The name resolver resolves a ResolvedAddresses with an XdsConfig without balancerName
+        //    field. Use the bootstrap information to create a channel.
+        // 3. Non EDS-only usecase:
+        //    XDS_CLIENT_REF attribute is available from ResolvedAddresses either from
+        //    XdsNameResolver or CDS policy.
+        // We assume XdsConfig switching happens only within one usecase, and there is no switching
+        // between different usecases.
+        if (xdsConfig.balancerName != null) {
+          // This is the EDS-only legacy usecase that requires balancerName from xds config.
+
+          if (oldXdsConfig == null || !xdsConfig.balancerName.equals(oldXdsConfig.balancerName)) {
+            final ManagedChannel channel = initLbChannel(
+                helper, xdsConfig.balancerName, Collections.<ChannelCreds>emptyList());
+            final Node node = Node.newBuilder()
+                .setMetadata(Struct.newBuilder()
+                    .putFields(
+                        "endpoints_required",
+                        Value.newBuilder().setBoolValue(true).build()))
+                .build();
+            // TODO(zdapeng): Use XdsClient to do Lrs directly.
+            lrsClient = loadReportClientFactory.createLoadReportClient(
+                channel, helper, new ExponentialBackoffPolicy.Provider(), loadStatsStore);
```

Comment:
Shouldn't `ExponentialBackoffPolicy.Provider` be passed down from the very first caller (which is `XdsLoadBalancer`)? Although we are not doing end-to-end test for `XdsLoadBalancer`, it would still be elegant to grab resources from a single source. (Notice you are creating two `ExponentialBackoffPolicy.Provider` here, one for LRS, one for `XdsComms`).

Reply:
The very first caller `XdsLoadBalancer2` as well as `LookasideLb` do not have `ExponentialBackoffPolicy.Provider` in its constructor, and `XdsLoadBalancer2` does not have any code about backoff, about `LoadBalancerRegistry`, or about `Random`, there are so many similar things we inject in component classes for testing. As long as we don't have end to end test from `XdsLoadBalancer2` on these particular behaviors, adding them would be too much for a class passing constructor arguments all the way down to its component classes and only used for them not itself.

I prefer creating two `ExponentialBackoffPolicy.Provider` as they are independent and cheap. If we do want a singleton. `ExponentialBackoffPolicy.Provider` class should provide a singleton for consumers to use.

Reply:
Ok, not a big issue.